base/v0_5_exp/translate.go

```diff
@@ -218,6 +220,155 @@ func translateLink(from Link, options common.TranslateOptions) (to types.Link, t
 	return
 }
 
+func translatePasswdUser(from PasswdUser, options common.TranslateOptions) (to types.PasswdUser, tm translate.TranslationSet, r report.Report) {
+	tr := translate.NewTranslator("yaml", "json", options)
+	tm, r = translate.Prefixed(tr, "gecos", &from.Gecos, &to.Gecos)
+	translate.MergeP(tr, tm, &r, "groups", &from.Groups, &to.Groups)
+	translate.MergeP2(tr, tm, &r, "home_dir", &from.HomeDir, "homeDir", &to.HomeDir)
+	translate.MergeP(tr, tm, &r, "name", &from.Name, &to.Name)
+	translate.MergeP2(tr, tm, &r, "no_create_home", &from.NoCreateHome, "noCreateHome", &to.NoCreateHome)
+	translate.MergeP2(tr, tm, &r, "no_log_init", &from.NoLogInit, "noLogInit", &to.NoLogInit)
+	translate.MergeP2(tr, tm, &r, "no_user_group", &from.NoUserGroup, "noUserGroup", &to.NoUserGroup)
+	translate.MergeP2(tr, tm, &r, "password_hash", &from.PasswordHash, "passwordHash", &to.PasswordHash)
+	translate.MergeP2(tr, tm, &r, "primary_group", &from.PrimaryGroup, "primaryGroup", &to.PrimaryGroup)
+	translate.MergeP2(tr, tm, &r, "should_exist", &from.ShouldExist, "shouldExist", &to.ShouldExist)
+	translate.MergeP(tr, tm, &r, "shell", &from.Shell, &to.Shell)
+	translate.MergeP(tr, tm, &r, "system", &from.System, &to.System)
+	translate.MergeP(tr, tm, &r, "uid", &from.UID, &to.UID)
+
+	var c path.ContextPath
+	if from.SSHAuthorizedKeys != nil && len(from.SSHAuthorizedKeys) > 0 {
+		c = path.New("yaml", "ssh_authorized_keys")
+		tm.AddTranslation(c, path.New("json", "sshAuthorizedKeys"))
+		for i, sshKey := range from.SSHAuthorizedKeys {
+			sshKey = SSHAuthorizedKey(strings.TrimSpace(string(sshKey)))
+			if len(sshKey) == 0 {
+				r.AddOnError(c, common.ErrSSHKeyEmpty)
+			}
+			tm.AddTranslation(c, path.New("json", fmt.Sprintf("sshAuthorizedKeys.%d", i)))
+			to.SSHAuthorizedKeys = append(to.SSHAuthorizedKeys, types.SSHAuthorizedKey(sshKey))
+		}
+	}
+
+	if from.SSHAuthorizedKeysLocal != nil && len(from.SSHAuthorizedKeysLocal) > 0 {
+		c = path.New("yaml", "ssh_authorized_keys_local")
+		tm.AddTranslation(c, path.New("json", "sshAuthorizedKeys"))
+		for _, sshKeyFile := range from.SSHAuthorizedKeysLocal {
+			if options.FilesDir == "" {
+				r.AddOnError(c, common.ErrNoFilesDir)
+				return
+			}
+
+			// calculate file path within FilesDir and check for
+			// path traversal
+			filePath := filepath.Join(options.FilesDir, sshKeyFile)
+			if err := baseutil.EnsurePathWithinFilesDir(filePath, options.FilesDir); err != nil {
+				r.AddOnError(c, err)
+				continue
+			}
+			contents, err := ioutil.ReadFile(filePath)
+			if err != nil {
+				r.AddOnError(c, err)
+				continue
+			}
+
+			sshKeyStrings := strings.TrimSpace(string(contents))
+			if len(contents) == 0 || len(sshKeyStrings) == 0 {
+				r.AddOnError(c, common.ErrFileEmpty)
```

Comment:
I'd be happy with a warning as well but if a user adds a blank file as source for SSH keys I'd at least like to make them aware of that because SSH keys can't be empty and they should double check their input – if they want to omit them, don't specify `ssh_authorized_keys(_local)`.

Reply:
Well, users might want to use a static Butane config but dynamically generate a list of zero or more SSH keys.  It's awkward to have a warning in that case.  Is there reason to expect that users will correctly create the file, but fail to put anything in it?

Reply:
I am not going to speculate on how users will actually use this feature and what is reasonable user behaviour – I'll be wrong anyway ;-). I'll get rid of this and we can discuss technical and semantical validation of inputs separately.